### DISCUSSION TITLE
Multi level Kepler Deployment

### DISCUSSION THREAD

**Discussion by tawalaya** [sustainable-computing-io/kepler#1209] [Q&A]
Hi there,
I'm looking into deploying Kepler in a KVM cluster that will host multiple Kubernetes nodes. Could you give me any guidance on whether this is possible and, if so, how to do it? 

In the documentation, the possibility to do so is hinted at. (VM with node info and power passthrough from BM ) I'm guessing that the RPM package needs to be installed on the KVM hosts, and the VMs need to mount specific files. But I couldn't find any precise doc on that.

Thanks in advance :)

**Comment by rootfs** [comment]
Kepler can run inside of VMs and use ML models to estimate power consumption. The VMs don't have to have VM passthrough (actually the RAPL counters won't be on VMs anyway).

**Comment by tawalaya** [comment]
Yes, I understand, but my question was. What if I wanted to use passthrough? I've seen models like [scaphandre](https://github.com/hubblo-org/scaphandre) where you can pass a custom rapl counter through to the VM that only accounts for the usage of the VM. Is something like this possible? Alternatively, what files is Kepler looking for to read RAPL counters? Could a combination scaphandre BM + VM Kepler + Kubernetes work?

**Reply by rootfs** [reply]
I am not sure if the mainstream KVM has already supported RAPL yet. There is [patch](https://lore.kernel.org/all/20240125072214.318382-4-aharivel@redhat.com/) authored by @aharivel 

If your VM is configured with host passthrough and has access to a virtualized RAPL counter, Kepler will use the counter to get energy readings, instead of using ML models to estimate.

**Reply by aharivel** [reply]
thanks @rootfs for pointing out my patch. 

Indeed there is no way at the moment to mount the RAPL driver inside a VM with QEMU/KVM. 
The patch, I'm working on, is exactly trying to solve this issue, without any passthrough. 
It's still a WIP, the biggest issue I'm facing is this [CVE](https://www.intel.com/content/www/us/en/developer/articles/technical/software-security-guidance/advisory-guidance/running-average-power-limit-energy-reporting.html#cve-2020-8694) because it involves security and no one want to open a breach in the virtualization stack!

**Reply by tiwatsuka** [reply]
Has the exploration of passthrough approach been stopped? If I understand correctly, Kepler on VM can obtain estimated power metrics from Kepler on BM without reading RAPL counter.
https://sustainable-computing.io/usage/deep_dive/

**Reply by aharivel** [reply]
@tiwatsuka Actually it has landed in Qemu just yesterday: 
https://gitlab.com/qemu-project/qemu/-/commit/0418f90809aea5b375c859e744c8e8610e9be446

**Reply by tiwatsuka** [reply]
Thank you @aharivel! It's great that Kepler on VM can obtain components power from RAPL. 

Having said that, I believe multi level Kepler deployment still useful to collect other estimated VM power metric such as platform power from ACPI or Redfish. I know that I can use power estimation models, but I prefer measurement approach in terms of  its explainability and accuracy without preparation.

**Reply by aharivel** [reply]
@tiwatsuka I understand your concern and this kind of decision is out of my hand. 

However, in my opinion, getting power measurement from IPMI or ACPI is not really handy for what Kepler tries to achieve. 
Kepler is not just here to export raw data, it processes them before exporting them. RAPL has the smallest granularity in terms of power plane (Package, Core, Dram,...) and it's the most convenient for calculating the ratio of each processes. 
Using platform power would include power usage of disks, fans, others hardware etc. and this is not really possible to know exactly which processes has triggered the fan, why the disk are spinning more at this points, and so on. I guess you get my point.

In the end, if you need to have the global power consumption of your system to add more visibility of your power usage, adding a IPMI / Redfish prometheus exporter is just as easy as running Kepler. However if you need to target which processes is the most power hungry, Kepler is way more convenient for that matter.

**Reply by tiwatsuka** [reply]
@aharivel I understand your point clearly. My concerns may stem from my specific use case. I am attempting to use Kepler not only to understand the trend of energy consumption but also to calculate the greenhouse gas emissions of each pod, in accordance with standards such as the GHG Protocol. I’m uncertain whether the electricity consumed by components like fans or disks can be disregarded even if it’s unclear which process is responsible for this consumption. Therefore, it would be beneficial for me if the estimated total energy consumption of each pod on VM could be presented in an explainable manner.

**Reply by aharivel** [reply]
@tiwatsuka Did you already have a look at this blog:
https://deploy-preview-435--tag-env-sustainability.netlify.app/blog/2024-06-idle-power-matters-kepler-metrics-for-public-cloud-energy-efficiency/
Maybe you could find answer in it ?

**Reply by imraneajel** [reply]
@aharivel  Thank you for your efforts !
How can I configure VM passthrough so that Kepler (deployed in Kubernetes inside the VM) can access RAPL and expose real‑time power measurements? 

And also, is there a way to configure WSL to access RAPL ? 
Thank you !

**Reply by aharivel** [reply]
@imraneajel If your system uses QEMU/KVM for virtualization, you can use that:  https://www.qemu.org/docs/master/specs/rapl-msr.html

Otherwise, I've got no idea, sorry.

**Reply by iajellabi** [reply]
@aharivel Thank you for your reply !
I use QEMU/KVM in my system, but how can I configure Kepler to read into the vMSRs in the VM ? 
as there is no powercap, RAPL power sources, Kepler can't give real-time measurements but an ML estimation.

**Reply by aharivel** [reply]
@iajellabi Please note that the vMSR is an experimental feature in QEMU. There is no warranty in any kind using it. However if you want to experiment with it follow:    https://www.qemu.org/docs/master/specs/rapl-msr.html#usage

Then inside the VM using it, the folder `/sys/class/powercap/intel-rapl/` should be populated and then for example "/sys/class/powercap/intel-rapl/intel-rapl:0/energy_uj" would be updating like the bare metal version. 

Then inside the VM you can run Kepler normally and it will give you the power metrcis. Again, this is all experimental.

**Comment by imraneajel** [comment]
Did you find the solution to do multi level deployment ? if yes, can you clarify on details how you did it ? Thank you !
